src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs

```diff
@@ -0,0 +1,295 @@
+                var earnings = earningsByLearner.EarningsForLearner(uln);
+
+                foreach (var earning in earnings)
+                {
+                    if (earning.HasNonIncentiveEarnings())
+                    {
+                        ProcessEarning(accountsWithNonPayableFlagSet, earning, 
+                            learnerCommitments, result, TransactionTypesFlag.AllLearning);
+                    }
+
+                    if (earning.HasFirstIncentive())
+                    {
+                        ProcessEarning(accountsWithNonPayableFlagSet, earning,
+                            learnerCommitments, result, TransactionTypesFlag.FirstEmployerProviderIncentives);
+                    }
+
+                    if (earning.HasSecondIncentive())
+                    {
+                        ProcessEarning(accountsWithNonPayableFlagSet, earning,
+                            learnerCommitments, result, TransactionTypesFlag.SecondEmployerProviderIncentives);
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private void ProcessEarning(ImmutableHashSet<long> accountsWithNonPayableFlagSet, RawEarning earning,
+            LearnerCommitments learnerCommitments, DatalockValidationResult result, TransactionTypesFlag earningType)
+        {
+            var censusDate = CalculateCensusDate(earning, earningType);
+            var commitments = learnerCommitments.ActiveCommitmentsForDate(censusDate).ToList();
+            var matchResult = _datalockMatcher.Match(commitments, earning, censusDate);
+            ValidateInitialResult(earning, matchResult.ErrorCodes, earningType,
+                commitments, result, accountsWithNonPayableFlagSet, learnerCommitments.Commitments, censusDate);
+        }
+
+        private DateTime CalculateCensusDate(RawEarning earning, TransactionTypesFlag earningType)
+        {
+            var date = new DateTime(1900, 01, 01);
+            switch (earningType)
+            {
+                case TransactionTypesFlag.AllLearning:
+                    date = CalculateOnProgCensusDate(earning);
+                    break;
+                case TransactionTypesFlag.FirstEmployerProviderIncentives:
+                    date = earning.FirstIncentiveCensusDate ?? new DateTime(1900,01,01);
+                    break;
+                case TransactionTypesFlag.SecondEmployerProviderIncentives:
+                    date = earning.SecondIncentiveCensusDate ?? new DateTime(1900, 01, 01);
+                    break;
+            }
+
+            return date;
+        }
+
+        public void ValidateInitialResult(RawEarning earning, List<string> errors, TransactionTypesFlag paymentType,
+            List<Commitment> commitments, DatalockValidationResult result, ImmutableHashSet<long> accountsWithNonPayableFlagSet,
+            List<Commitment> allCommitments, DateTime censusDate)
+        {
```

Comment:
It's quite unclear what this function is actually doing? It's either updating the errors parameter or adding a record to DataLockValidationResult parameter.